packages/react/modules/components/configure-adapter/configure-adapter.js

```diff
@@ -27,19 +23,96 @@ type Props = {
   defaultFlags?: Flags,
   children: React.Component<any>,
 };
+type State = {
+  adapterArgs: AdapterArgs,
+};
 type AdapterState = $Values<typeof AdapterStates>;
+type AdapterReconfigurationOptions = {
+  exact?: boolean,
+};
+type AdapterReconfiguration = {
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions,
+};
+type ReconfigureAdapter = (
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions
+) => void;
+
+export const AdapterContext: Context<ReconfigureAdapter> = createReactContext(
+  () => {}
+);
 
-export default class ConfigureAdapter extends PureComponent<Props> {
+export const mergeAdapterArgs = (
+  previousAdapterArgs: AdapterArgs,
+  { adapterArgs: nextAdapterArgs, options }: AdapterReconfiguration
+): AdapterArgs =>
+  options.exact ? nextAdapterArgs : merge(previousAdapterArgs, nextAdapterArgs);
+
+export default class ConfigureAdapter extends PureComponent<Props, State> {
   static defaultProps = {
     shouldDeferAdapterConfiguration: false,
     children: null,
     defaultFlags: {},
   };
 
   adapterState: AdapterState = AdapterStates.UNCONFIGURED;
-  setAdapterState = (nextAdapterState: AdapterState): void => {
+  pendingAdapterArgs: ?AdapterArgs = null;
+
+  state: { adapterArgs: AdapterArgs } = {
+    adapterArgs: this.props.adapterArgs,
+  };
+
+  setAdapterState = (
+    nextAdapterState: AdapterState,
+    exact: boolean = false
+  ): void => {
     this.adapterState = nextAdapterState;
   };
+  setAdapterArgs = (nextAdapterArgs: AdapterArgs): void =>
+    this.setState(prevState => ({
+      ...prevState,
+      adapterArgs: nextAdapterArgs,
+    }));
+
+  /**
+   * NOTE:
+   *   This is passed through the React context (it's a public API).
+   *   Internally this component has a `ReconfigureAdapter` type;
+   *   this function has two arguments for clarify.
+   */
+  reconfigureOrQueue = (
+    nextAdapterArgs: AdapterArgs,
+    options: AdapterReconfigurationOptions = {}
+  ): void =>
+    this.adapterState === AdapterStates.CONFIGURED &&
+    this.adapterState !== AdapterStates.CONFIGURING
+      ? this.setAdapterArgs(
+          mergeAdapterArgs(this.state.adapterArgs, {
+            adapterArgs: nextAdapterArgs,
+            options,
+          })
+        )
+      : this.setPendingAdapterArgs({ adapterArgs: nextAdapterArgs, options });
+
+  setPendingAdapterArgs = (
+    nextReconfiguration: AdapterReconfiguration
+  ): void => {
+    /**
+     * NOTE:
+     *    The next reconfiguration is merged into the previous
+     *    one instead of maintaining a queue.
+     */
+    this.pendingAdapterArgs = mergeAdapterArgs(
+      this.pendingAdapterArgs || this.state.adapterArgs,
```

Comment:
Hmm, can you clarify this choice?

Reply:
Sure. Imagine you have two `ReconfigureFlopflip` components somewhere flying around you component tree. One with `user={{ firstRun: 'foo' }}` the second with `user={{ secondRun: 'bar' }}`. Both may render (trigger reconfiguration) before the adapter even initialized. Given that the first queues a reconfiguration the `pendingAdapterArgs` are populated. Given that after the second ran we want to reconfigure with merged `adapterArgs` of both we have to `mergeAdapterArgs` with the `pendingsAdapterArgs` as using the `state.appliedAdapterArgs` would imply the that lastly rendered `ReconfigureFlopflip` always wins. Hope that makes sense.

Reply:
Hmm thanks for the explanation. I’m still a bit confused though.
For example, can’t we check if pendingAdapterArgs is defined and only then do the merge with the state.appliedAdapterArgs?

Also, what would happen if:
* first reconfigure renders (Adapter not ready)
* Adapter ready
* Second reconfigure renders

Reply:
I think I've a knot in my brain. We have to merge

1. With `pendingAdapterArgs` whenever multiple calls to `reconfigureOrQueue` are performed while the adapter was still configuring. To accumulate the `pendingAdapterArgs` (merged instead of maintaining a queue).
2. With `appliedAdapterArgs` whenever the adapter is configured and is able to "flush" the `nextAdapterArgs` straight away.

The first time round we merge with `appliedAdapterArgs` (from `componentDidMount`) as `nextAdapterArgs` passed to `reconfigureOrQueue` might be only "shallow".

To you example:

1. Adapter not ready: `reconfigureOrQueue` is called -> `pendingAdapterArgs` are a deep merge of `appliedAdapterArgs` and the passed `nextAdapterArgs`
2. Adapter ready: the `pendingAdapterArgs` are applied through `unsetPendingAdapterArgs` and "flushed" through `setAdapterArgs` (which could be renamed to `applyAdapterArgs`
3. Adapter still ready: Second call to `reconfigureOrQueue` -> No call to `setPendingAdapterArgs`, straight to `setAdapterArgs` which flushes them straight away.

Hope I make sense. Please correct me if I am wrong.